runtime/commonMain/src/kotlinx/serialization/Annotations.kt

```diff
@@ -10,60 +10,156 @@ import kotlinx.serialization.json.*
  */
 @Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Serializer(
-    val forClass: KClass<*> // what class to create serializer for
+    val forClass: KClass<*> //  target class to create serializer for
 )
 
 /**
- * Overrides name visible to the runtime part of serialization framework
+ * Overrides the name of the current element in the [SerialDescriptor].
+ *
+ * By default, [SerialDescriptor.serialName] and [SerialDescriptor.getElementName]
+ * are matched with fully-qualified name of the target class and the name of the property respectively.
+ * Applying this annotation changes the visible name to the given [value]:
+ *
+ * ```
+ * package foo
+ *
+ * @Serializable
+ * class Foo(val myInt: Int)
+ *
+ * @Serializable
+ * @SerialName("Bar")
+ * class Bar(@SerialName("int") val myInt: Int)
+ *
+ * fun printNames(descriptor: SerialDescriptor) {
+ *     println(descriptor.serialName)
+ *     println(descriptor.getElementName(0))
+ * }
+ *
+ * // Prints "foo.Foo" and "myInt"
+ * printNames(Foo.serializer().descriptor)
+ *
+ * // Prints "Bar" and "int"
+ * printNames(Bar.serializer().descriptor)
+ * ```
+ * Element names and serial names are used by text-based serial formats in order to encode the name of the class or
+ * the name of the property, e.g. by [Json].
  */
 @Target(AnnotationTarget.PROPERTY, AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
 public annotation class SerialName(val value: String)
 
 /**
- * Indicates that property must be present during deserialization process,
- * even if it has default value.
+ * Indicates that property must be present during deserialization process, despite having a default value.
  */
 @Target(AnnotationTarget.PROPERTY)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Required
 
 /**
- * Marks this property invisible for whole serialization framework.
- * Transient properties must have default values.
+ * Marks this property invisible for the whole serialization process, including [serial descriptors][SerialDescriptor].
+ * Transient properties should have default values.
  */
 @Target(AnnotationTarget.PROPERTY)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Transient
 
 /**
- * When annotation class is marked with `@SerialInfo`, compiler plugin can instantiate it
- * and put into [SerialDescriptor], to be retrieved later during serialization process.
+ * Meta-annotation that commands the compiler plugin to handle the annotation as serialization-specific.
+ * Serialization-specific annotations are preserved in the [SerialDescriptor] and can be retrieved
+ * during serialization process with [SerialDescriptor.getElementAnnotations]:
+ *
+ * ```
+ * /**
+ *  * Specifies ProtoBuf unique id for a field in the protobuf message, assigned to a Kotlin property
+ *  */
+ * @SerialInfo
+ * @Target(AnnotationTarget.PROPERTY)
```

Comment:
This example is not useful unless it's shown how to retrieve it.

I suggest removing example completely, since declaring serialinfo annotations is mainly for format authors. For them, we'll eventually have an advanced guide on custom formats.